rpcs3/Emu/Cell/PPUThread.cpp

```diff
@@ -368,52 +370,60 @@ extern bool ppu_patch(u32 addr, u32 value)
 	return true;
 }
 
-std::string ppu_thread::dump() const
+std::string ppu_thread::dump_all() const
 {
-	std::string ret = cpu_thread::dump();
-	fmt::append(ret, "Priority: %d\n", +prio);
-	fmt::append(ret, "Stack: 0x%x..0x%x\n", stack_addr, stack_addr + stack_size - 1);
-	fmt::append(ret, "Joiner: %s\n", joiner.load());
-	fmt::append(ret, "Commands: %u\n", cmd_queue.size());
+	std::string ret = cpu_thread::dump_misc() + '\n';
 
-	const char* _func = current_function;
+	ret += dump_misc() + '\n';
+	ret += dump_regs() + '\n';
+	ret += dump_callstack();
 
-	if (_func)
-	{
-		ret += "Current function: ";
-		ret += _func;
-		ret += '\n';
+	return ret;
+}
 
-		for (u32 i = 3; i <= 6; i++)
-			fmt::append(ret, " ** GPR[%d] = 0x%llx\n", i, syscall_args[i - 3]);
-	}
-	else if (is_paused())
+std::string ppu_thread::dump_regs() const
+{
+	std::string ret;
+
+	for (uint i = 0; i < 32; ++i)
 	{
-		if (const auto last_func = last_function)
+		auto reg = gpr[i];
+
+		fmt::append(ret, "GPR[%-2d] = 0x%-8llx", i, reg);
+
+		const u32 max_str_len = 32;
+		const u32 hex_count = 10;
+
+		if (vm::check_addr(reg, max_str_len, vm::page_readable))
 		{
-			_func = last_func;
-			ret += "Last function: ";
-			ret += _func;
-			ret += '\n';
-		}
-	}
+			const u64 reg_ptr = vm::read64(reg);
 
-	if (const auto _time = start_time)
-	{
-		fmt::append(ret, "Waiting: %fs\n", (get_guest_system_time() - _time) / 1000000.);
-	}
-	else
-	{
-		ret += '\n';
-	}
+			if (vm::check_addr(reg_ptr, max_str_len, vm::page_readable))
+			{
+				reg = reg_ptr;
+			}
 
-	if (!_func)
-	{
-		ret += '\n';
-	}
+			const auto gpr_buf = vm::get_super_ptr<u8>(reg);
+
+			std::string buf_tmp(gpr_buf, gpr_buf + max_str_len);
 
-	ret += "\nRegisters:\n=========\n";
-	for (uint i = 0; i < 32; ++i) fmt::append(ret, "GPR[%d] = 0x%llx\n", i, gpr[i]);
+			if (std::isprint(buf_tmp[0]) && std::isprint(buf_tmp[1]) && std::isprint(buf_tmp[2]))
```

Comment:
There is UB here, see https://en.cppreference.com/w/cpp/string/byte/isspace for exaplenation.
I suggest the following change
```suggestion
			if (std::isprint(buf_tmp[0] + 0u) && std::isprint(buf_tmp[1] + 0u) && std::isprint(buf_tmp[2] + 0u))
```